docker/types/daemon.py

```diff
@@ -0,0 +1,72 @@
+from socket import SHUT_RDWR
+
+try:
+    import requests.packages.urllib3 as urllib3
+except ImportError:
+    import urllib3
+
+
+class CancellableStream(object):
+    """
+    Stream wrapper for real-time events, logs, etc. from the server.
+
+    Example:
+        >>> events = client.events(cancellable=True)
+        >>> for event in events.stream():
+        ...   print event
+        >>> # and cancel from another thread
+        >>> events.close()
+    """
+
+    def __init__(self, stream, response):
+        self._stream = stream
+        self._response = response
+
+    def stream(self):
+        """
+        Start streaming the events from the server.
+
+        Returns:
+            (generator): A blocking generator you can iterate over to retrieve
+                events as they happen.
+
+        Raises:
+            The error from the underlying generator, except for
+                ``urllib3.exceptions.ProtocolError`` which is expected to
+                happen when the stream is closed.
+        """
+
+        try:
+            for item in self._stream:
+                yield item
+        except urllib3.exceptions.ProtocolError:
+            pass
+
+    def __iter__(self):
+        return self.stream()
+
+    def close(self):
+        """
+        Closes the event streaming.
+        """
+
```

Comment:
Can't we simply call `self._response.close()`? http://docs.python-requests.org/en/master/api/#requests.Response.close

Reply:
No, unfortunately, it doesn't interrupt the blocking socket read.
From what I could find out, it looks like we need the `socket.shutdown(SHUT_RDWR)` call to do that fo us.